lisp/ess-s-lang.el

```diff
@@ -575,68 +575,49 @@ and one that is well formatted in emacs ess-mode."
-R/S, the underscore is always inserted.
-
-If `ess-smart-S-assign-key' is nil, just call
-`self-insert-command'. You can pass N as usual."
+(defun ess-cycle-assignment ()
+  "Cycle between assignment symbols in `ess-assign-list'.
+On consecutive calls, replace the assignment symbol before point
+with the next symbol from that list."
   (interactive)
-  (if ess-smart-S-assign-key
-      (save-restriction
-        (ignore-errors
-          (when (and (eq major-mode 'inferior-ess-mode)
-                     (> (point) (process-mark (get-buffer-process (current-buffer)))))
-            (narrow-to-region (process-mark (ess-get-process)) (point-max)))
-          (and ess-noweb-mode
-               (ess-noweb-in-code-chunk)
-               (ess-noweb-narrow-to-chunk))
-          (and (fboundp 'pm/narrow-to-span)
-               polymode-mode
-               (pm/narrow-to-span)))
-        (if (or
-             (ess-inside-string-or-comment-p (point))
-             (not (equal ess-language "S")))
-            (insert ess-smart-S-assign-key)
-          (ess-insert-S-assign)))
-    (funcall #'self-insert-command (or N 1))))
+  (if (eq last-command this-command)
+      (let ((slist ess-assign-list)
+            str)
+        (while (and (setq str (car slist)
+                          slist (cdr slist))
+                    (not (and (re-search-backward str
+                                                  (- (point) (length str)) t)
+                              (not (replace-match (car slist))))))))
+    (insert (car ess-assign-list))))
 
 (defun ess-insert-S-assign ()
-  "Insert the assignment operator `ess-S-assign', unless it is already there.
-In that case, it is removed and replaced by `ess-smart-S-assign-key'.
-  `ess-S-assign', typically \" <- \", can be customized."
+  "Insert the first element of `ess-assign-list'.
+When `ess-maybe-replace-S-assign' is non-nil, if the character
+before point is the first element of `ess-assign-list', replace
+it with the character typed if it is an ASCII character,
+otherwise remove the assignment."
   (interactive)
-  ;; one keypress produces ess-S-assign; a second keypress will delete
-  ;; ess-S-assign and instead insert _
-  ;; Rather than trying to count a second _ keypress, just check whether
-  ;; the current point is preceded by ess-S-assign.
-  (let ((assign-len (length ess-S-assign)))
-    (if (and
-         (>= (point) (+ assign-len (point-min))) ;check that we can move back
-         (save-excursion
-           (backward-char assign-len)
-           (looking-at ess-S-assign)))
-        ;; If we are currently looking at ess-S-assign, replace it with _
-        (progn
-          (delete-char (- assign-len))
-          (insert ess-smart-S-assign-key))
-      (if (string= ess-smart-S-assign-key "_")
-          (delete-horizontal-space))
-      (insert ess-S-assign))))
+  (let ((assign (car ess-assign-list))
+        (event (and (numberp last-input-event)
```

Comment:
Use `simple-event-type` here.

Reply:
`simple-event-type` doesn't appear to be defined on my Emacs. Do you mean `event-basic-type`?

Reply:
Yes. This is what I meant.